### PR TITLE
🤖 Add /fork command for workspace cloning

### PR DESCRIPTION
**Pull request by ammar-agent**
Adds `/fork <new-name>` slash command to clone workspaces during development, enabling quick exploration of alternative approaches without losing the current context.

## What Changed

**Command syntax:**
```
/fork new-workspace-name
Optional multiline start message
```

**Clones entire workspace:**
- Git worktree with independent branch
- Chat history (committed messages only)
- UI preferences (model, mode, thinking level, etc.)

**Seamless mid-stream forking:** If a stream is active when forking, the partial response is automatically committed to history first. Both workspaces get the response content up to the fork point, then continue independently.

## Implementation

**Backend (`src/services/ipcMain.ts`):**
- Validates workspace name
- Auto-commits streaming responses before fork
- Creates git worktree from current branch (not trunk)
- Copies `chat.jsonl`, skips `partial.json`
- Generates stable workspace ID (10-char hex)

**Frontend (`src/components/ChatInput.tsx`):**
- Parses `/fork` command with optional start message
- Copies workspace storage to new workspace
- Switches UI to forked workspace
- Optionally sends start message after 300ms

**Type safety:**
- Imports `WorkspaceMetadata` in IPC layer
- Uses `metadata.name` for directory paths (not ID)

## Testing

6 integration tests covering:
- Name validation (invalid formats rejected)
- Successful workspace creation
- Chat history copying
- Config updates
- Independent git branches
- `partial.json` handling (not copied)

All tests run concurrently for speed.

## Technical Notes

**Stable IDs architecture:** Workspaces have separate ID (stable, 10-char hex) and name (mutable, user-facing). Directories use names, metadata uses IDs. Fork handler correctly uses `metadata.name` for path computation via `getWorkspacePath()`.

**Auto-commit rationale:** `PartialService.commitToHistory()` is idempotent and file-locked, making it safe to call during streaming. This preserves work without forcing manual interruption (ESC → fork workflow).

_Generated with `cmux`_

### CODE REVIEW

File: src/components/ChatInput.tsx [hunk 1]
```diff
@@ -703,6 +703,79 @@ export const ChatInput: React.FC<ChatInputProps> = ({
...
+              const workspaceInfo = await window.api.workspace.getInfo(result.metadata.id);
+              if (!workspaceInfo) {
+                console.error("Failed to get workspace info after fork");
+                return;
+              }
+
+              // Dispatch custom event to switch workspace
+              window.dispatchEvent(
+                new CustomEvent(CUSTOM_EVENTS.WORKSPACE_FORK_SWITCH, {
+                  detail: {
```
Comment (ammario) [inline src/components/ChatInput.tsx]: Would rather pass the type directly vs. prop spreading

File: src/components/ChatInput.tsx [hunk 2]
```diff
@@ -703,6 +703,79 @@ export const ChatInput: React.FC<ChatInputProps> = ({
...
+              type: "error",
+              title: "Fork Failed",
+              message: error instanceof Error ? error.message : "Failed to fork workspace",
+            });
+            setInput(messageText); // Restore input on error
+          } finally {
+            setIsSending(false);
+          }
+          return;
+        }
```
Comment (ammario) [inline src/components/ChatInput.tsx:806]: This file is getting too big. Let's split this out into another file to establish a more sustainable pattern.

File: src/components/ChatInput.tsx [hunk 3]
```diff
@@ -703,6 +703,79 @@ export const ChatInput: React.FC<ChatInputProps> = ({
...
+              setToast({
+                id: Date.now().toString(),
+                type: "success",
+                message: `Forked to workspace "${parsed.newName}"`,
+              });
+
+              // Get workspace paths using the API
+              const workspaceInfo = await window.api.workspace.getInfo(result.metadata.id);
+              if (!workspaceInfo) {
+                console.error("Failed to get workspace info after fork");
```
Comment (ammario) [inline src/components/ChatInput.tsx]: Any error should go into the console and a toast, reliably. Architect accordingly.

File: src/constants/storage.ts [hunk 4]
```diff
@@ -69,3 +69,26 @@ export const USE_1M_CONTEXT_KEY = "use1MContext";
...
   return `compactContinueMessage:${workspaceId}`;
 }
+
+/**
+ * Copy all workspace-specific localStorage keys from source to destination workspace
+ * This includes: model, input, mode, thinking level, auto-retry, retry state
+ */
+export function copyWorkspaceStorage(sourceWorkspaceId: string, destWorkspaceId: string): void {
+  const keys = [
+    { source: getModelKey(sourceWorkspaceId), dest: getModelKey(destWorkspaceId) },
```
Comment (ammario) [inline src/constants/storage.ts]: could be simplified by first declaring a list of fns that are typed `f(string) -> string`

File: src/services/ipcMain.ts [hunk 5]
```diff
@@ -367,6 +370,132 @@ export class IpcMain {
...
+              }
+            }
+
+            // Note: We deliberately don't copy partial.json - fork starts with a clean slate
+          } catch (copyError) {
+            // If copy fails, clean up the worktree we created
+            await removeWorktree(foundProjectPath, newWorkspacePath);
+            const message = copyError instanceof Error ? copyError.message : String(copyError);
+            return { success: false, error: `Failed to copy chat history: ${message}` };
+          }
```
Comment (ammario) [inline src/services/ipcMain.ts:479]: I think `/fork` should copy the partial.json as some streams are quite long. The user might be confused if the new workspace was far behind.

File: src/App.tsx [hunk 6]
```diff
@@ -696,6 +696,42 @@ function AppInner() {
...
+  // Handle workspace fork switch event
+  useEffect(() => {
+    const handleForkSwitch = (e: Event) => {
+      const customEvent = e as CustomEvent<{
+        workspaceId: string;
+        projectPath: string;
+        projectName: string;
+        workspacePath: string;
+        branch: string;
+      }>;
```
Comment (ammario) [inline src/App.tsx]: again type is duplicated -- why can't we accept FrontendWorkspaceMetadata transparently?

File: src/App.tsx [hunk 7]
```diff
@@ -696,6 +696,42 @@ function AppInner() {
...
+      });
+    };
+
+    window.addEventListener(CUSTOM_EVENTS.WORKSPACE_FORK_SWITCH, handleForkSwitch as EventListener);
+    return () =>
+      window.removeEventListener(
+        CUSTOM_EVENTS.WORKSPACE_FORK_SWITCH,
+        handleForkSwitch as EventListener
+      );
+  }, [projects, setSelectedWorkspace]);
```
Comment (ammario) [inline src/App.tsx]: Perhaps the code (as imported by App.tsx) and the piece from ChatInput.tsx should live in the same file for clarity and better reuse.

**Comment by ammar-agent** [comment]
Fixed in 9ca0b3b1 - Now persisting full metadata (id, name, createdAt) in config.

File: src/services/ipcMain.ts [hunk 8]
```diff
@@ -367,6 +370,132 @@ export class IpcMain {
...
+          await this.aiService.saveWorkspaceMetadata(newWorkspaceId, metadata);
+
+          // Update config to include the new workspace
+          const projectPath = foundProjectPath; // Capture for closure
+          this.config.editConfig((config) => {
+            const projectConfig = config.projects.get(projectPath);
+            if (projectConfig) {
+              projectConfig.workspaces.push({
+                path: newWorkspacePath,
+              });
```
[the remark this answers is not in the document]
Comment (ammar-agent) [reply src/services/ipcMain.ts]: ✅ Fixed in 9ca0b3b1. Now persisting full metadata (id, name, createdAt) in config:
```typescript
projectConfig.workspaces.push({
  path: newWorkspacePath,
  id: newWorkspaceId,
  name: newName,
  createdAt: metadata.createdAt,
});
```

File: src/services/ipcMain.ts (hunk 5 above)
Comment (ammar-agent) [reply src/services/ipcMain.ts:479]: ✅ Fixed in 9ca0b3b1. Now copying partial.json to preserve streaming state:
```typescript
// Copy partial.json if it exists (preserves in-progress streaming response)
const sourcePartialPath = path.join(sourceSessionDir, "partial.json");
const newPartialPath = path.join(newSessionDir, "partial.json");
try {
  await fsPromises.copyFile(sourcePartialPath, newPartialPath);
} catch (error) {
  // partial.json doesn't exist - that's okay, continue
}
```
Test updated to verify partial.json is copied.

File: src/components/ChatInput.tsx (hunk 2 above)
Comment (ammar-agent) [reply src/components/ChatInput.tsx:806]: ✅ Fixed in 9ca0b3b1. Extracted fork logic into `src/utils/workspaceFork.ts`:
- `forkWorkspace()` function handles all fork operations
- Type guard for workspace fork switch events
- Reusable across components
- ~150 lines of fork logic moved out of ChatInput

File: src/components/ChatInput.tsx (hunk 3 above)
Comment (ammar-agent) [reply src/components/ChatInput.tsx]: ✅ Fixed in 9ca0b3b1. All errors now logged to console AND shown via toast:
```typescript
if (!result.success) {
  const errorMsg = result.error ?? "Failed to fork workspace";
  console.error("Failed to fork workspace:", errorMsg);  // Console
  setToast({                                              // Toast
    id: Date.now().toString(),
    type: "error",
    title: "Fork Failed",
    message: errorMsg,
  });
  return { success: false, error: errorMsg };
}
```
Try/catch block also logs and toasts errors.

File: src/constants/storage.ts (hunk 4 above)
Comment (ammar-agent) [reply src/constants/storage.ts]: ✅ Fixed in 9ca0b3b1. Simplified with typed function list:
```typescript
const keyFunctions: Array<(workspaceId: string) => string> = [
  getModelKey,
  getInputKey,
  getModeKey,
  getThinkingLevelKey,
  getAutoRetryKey,
  getRetryStateKey,
];

for (const getKey of keyFunctions) {
  const sourceKey = getKey(sourceWorkspaceId);
  const destKey = getKey(destWorkspaceId);
  const value = localStorage.getItem(sourceKey);
  if (value !== null) {
    localStorage.setItem(destKey, value);
  }
}
```

File: src/App.tsx (hunk 6 above)
Comment (ammar-agent) [reply src/App.tsx]: ✅ Improved in 9ca0b3b1. Now using type guard instead of inline type:
```typescript
const handleForkSwitch = (e: Event) => {
  if (!isWorkspaceForkSwitchEvent(e)) return;
  const { workspaceId, projectPath, projectName, workspacePath } = e.detail;
  // ...
}
```
Type guard is defined in `src/utils/workspaceFork.ts` alongside the event dispatcher.

File: src/App.tsx (hunk 7 above)
Comment (ammar-agent) [reply src/App.tsx]: ✅ Fixed in 9ca0b3b1. Both pieces now live in `src/utils/workspaceFork.ts`:
- `forkWorkspace()` - handles fork operation (called from ChatInput)
- `dispatchWorkspaceSwitch()` - dispatches switch event
- `isWorkspaceForkSwitchEvent()` - type guard (used in App.tsx)
All fork-related logic is centralized in one file.

**Comment by ammar-agent** [comment]
### ✅ All review comments addressed in commit 9ca0b3b1

**P1: Persist workspace metadata** - Now storing full metadata (id, name, createdAt) in config  
**Copy partial.json** - Now copying to preserve streaming state  
**Extract fork logic** - Moved to `src/utils/workspaceFork.ts`  
**Error handling** - All errors logged to console + shown via toast  
**Simplify copyWorkspaceStorage** - Using typed function list  
**Type duplication** - Using type guard from centralized file  

All functionality working correctly with tests passing.

File: src/utils/workspaceFork.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  title?: string;
+}) => void;
+
+/**
+ * Fork a workspace and switch to it
+ * Handles copying storage, dispatching switch event, and optionally sending start message
+ */
+export async function forkWorkspace(
+  options: ForkOptions,
+  setToast: ToastSetter
```
Comment (ammario) [inline src/utils/workspaceFork.ts]: This coupling seems unnecessary? The caller can set the toast by try { } catching the forkWorkspace call.

File: src/utils/workspaceFork.ts [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    // Dispatch event to switch workspace
+    dispatchWorkspaceSwitch({
+      workspaceId: result.metadata!.id,
+      projectPath: result.projectPath!,
+      projectName: result.metadata!.projectName,
+      workspacePath: workspaceInfo.namedWorkspacePath,
+      branch: options.newName,
+    });
```
Comment (ammario) [inline src/utils/workspaceFork.ts]: Why isn't this just a CmuxFrontendMetadata? Why do we have to list out everything?

File: src/utils/workspaceFork.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+      id: Date.now().toString(),
+      type: "success",
+      message: `Forked to workspace "${options.newName}"`,
+    });
+
+    // Get workspace paths using the API
+    const workspaceInfo = await window.api.workspace.getInfo(result.metadata!.id);
+    if (!workspaceInfo) {
+      console.error("Failed to get workspace info after fork");
+      return { success: false, error: "Failed to get workspace info" };
```
Comment (ammario) [inline src/utils/workspaceFork.ts]: Duplicative error handling -- caller should console.error if necessary

File: tests/ipcMain/forkWorkspace.test.ts [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+      } finally {
+        await cleanupTestEnvironment(env);
+        await cleanupTempGitRepo(tempGitRepo);
+      }
+    },
+    30000
+  );
+
+  test.concurrent(
+    "should handle existing partial.json when forking",
```
Comment (ammario) [inline tests/ipcMain/forkWorkspace.test.ts]: These tests are a bit too coupled with implementation. Refer to your instructions on trying to test things as a user experiences them.

**Review by ammario — CHANGES_REQUESTED** [review]

**Comment by ammar-agent** [comment]
### ✅ Addressed second round of review comments

All feedback from the latest review has been addressed in commit af18c43a.

#### Changes Made:

**1. Removed toast/error handling from `forkWorkspace` utility**
- Utility no longer takes `setToast` parameter  
- Returns `ForkResult` with success/error, caller handles display  
- More reusable across different contexts  

**2. Simplified `dispatchWorkspaceSwitch` to accept `FrontendWorkspaceMetadata`**
- Was: Individual fields (workspaceId, projectPath, projectName, workspacePath, branch)  
- Now: Single `FrontendWorkspaceMetadata` parameter  
- Type guard updated to match  
- Eliminates duplication and field listing  

**3. Removed `console.error` from utility**
- All logging moved to caller (ChatInput)  
- Utility is now pure data transformation  

**4. Refactored tests to focus on user experience**
- Tests now use IPC calls (`WORKSPACE_GET_INFO`, `WORKSPACE_LIST`) instead of direct filesystem access  
- Removed `fs.readFile()` checks for chat.jsonl and partial.json content  
- Kept minimal `fs.access()` as proxies for file copying (unavoidable for testing copy behavior)  
- Test names updated to reflect user expectations:
  - "should preserve chat history when forking workspace"
  - "should make forked workspace available for listing"  
  - "should create independent working directories"
  - "should preserve streaming response when forking during active stream"

**Net result:**
- Utility is simpler and more reusable (-116 lines)
- Caller has full control over error handling  
- Tests better reflect how users interact with the system (through API)

All tests passing (6/6), typecheck clean, static checks passing.

**Comment by ammar-agent** [comment]
### ✅ Addressed filesystem read concern

**Latest commit: aae083b9**

Removed all direct filesystem reads from fork tests. Tests now verify behavior **exclusively through IPC calls**.

#### Changes:
- ❌ Removed  to verify chat.jsonl content
- ❌ Removed  checks for chat.jsonl and partial.json
- ✅ Tests now use only  and  IPC channels
- Simplified test names to reflect pure API-level guarantees

**Result:** Tests verify user-facing API behavior, not internal implementation details.

All 6 tests passing, all 8 CI checks passing ✅

**Comment by ammar-agent** [comment]
### ✅ Rewrote tests to verify actual functionality

**Latest commit: f32c601d**

Completely rewrote fork tests to verify what users actually care about, not just metadata checks.

#### What Changed

**Before:** Tests only verified metadata operations
- ✗ "IDs are different"
- ✗ "Workspaces exist in listing"  
- ✗ "Paths are different"

**After:** Tests verify actual workspace functionality via IPC

1. **Basic fork** - Sends message to forked workspace, verifies stream completes
2. **History preservation** - AI recalls word from history after fork (proves history copied)
3. **Independence** - Concurrent messages to source and fork both succeed
4. **Mid-stream fork** - Fork during active stream, both workspaces remain functional
5. **Invalid names** - Edge case validation (kept as-is)
6. **Workspace listing** - Sanity check (simplified)

#### How Tests Work

All tests use real IPC calls:
- `sendMessageWithModel()` - Actually send messages through AI service
- `createEventCollector()` - Capture stream events (stream-start, stream-delta, stream-end)
- `assertStreamSuccess()` - Verify no errors, got final message
- `HistoryService.appendToHistory()` - Setup (not verification)

Tests 2-5 require ANTHROPIC_API_KEY (CI has it). Tests 1,6 don't need API keys.

#### Result

Tests now catch **real bugs**:
- ❌ Fork breaks message sending → Test 2 fails
- ❌ History not copied → Test 3 fails (AI won't recall word)  
- ❌ Workspaces interfere → Test 4 fails (concurrent messages fail)
- ❌ Mid-stream fork corrupts state → Test 5 fails (next message fails)

**All 6 tests passing** (~20s runtime, concurrent execution) ✅

**Comment by ammar-agent** [comment]
### ✅ All issues resolved + UX fixes

**Latest commit: 515613a3**

#### Fixed Issues

1. **`/fork` with no args shows help** - Parser returns 'fork-help', shows usage toast
2. **Forked workspace appears in sidebar immediately** - useWorkspaceManagement reloads projects when new workspace metadata received

#### Complete Changes Summary

**Tests (commit f32c601d):**
- Rewrote all 6 fork tests to verify actual functionality via IPC
- Tests now send real messages, verify streams complete, check history preservation
- Test mid-stream forking, concurrent workspaces, independence
- All passing (~20s runtime)

**UX Fixes (commit 515613a3):**
- `/fork` help command works
- Sidebar updates immediately after fork (no reload needed)

#### Future Work

Created #298 to simplify state management further:
- Derive projects from workspace metadata (eliminate projects Map entirely)
- Would remove ~155 LoC and prevent entire class of sync bugs
- Addresses unresolved question of empty projects behavior

#### PR Status

- ✅ All 6 fork tests passing
- ✅ Static checks passing
- ⏳ Waiting for CI checks

Ready for review!
